### PR TITLE
Full flow for fee tracking and distributing

### PR DESCRIPTION
**Pull request by cuong-incognito**
- Support new pDex params
- Tracking fee when liquidity is updated or a trade is executed
- Mint pDEX genesis block and reward blocks

### CODE REVIEW

File: blockchain/pdex/params.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		PDEXRewardPoolPairsShare:        paramsState.PDEXRewardPoolPairsShare(),
+		StakingPoolsShare:               paramsState.StakingPoolsShare(),
+		StakingRewardTokens:             paramsState.StakingRewardTokens(),
+		MintNftRequireAmount:            paramsState.MintNftRequireAmount(),
+		MaxOrdersPerNft:                 paramsState.MaxOrdersPerNft(),
+	}
+}
+
+func (p *Params) Clone() *Params {
+	result := &Params{}
```
Comment (dat-incognito-org) [inline blockchain/pdex/params.go:39]: it seems many fields are missing from Clone(). Is this intended ?

Comment (cuong-incognito) [reply blockchain/pdex/params.go:39]: Miss lines command when merged PRs. I will fix it soon, thanks!

File: blockchain/pdex/v2utils/producer.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+			strconv.Itoa(int(shardID)),
+			status,
+			string(reqContentBytes),
+		}
+		return [][]string{inst}
+	}
+
+	// To store the keys in slice in sorted order
+	keys := make([]common.Hash, len(receivers))
+	i := 0
```
Comment (dat-incognito-org) [inline blockchain/pdex/v2utils/producer.go:142]: this `i` does not seem to be used after the `i++`

File: blockchain/pdex/v2utils/reserve.go [hunk 3]
```diff
@@ -233,3 +233,178 @@ func MaybeAcceptTrade(amountIn, fee uint64, tradePath []string, receiver privacy
...
+	}
+
+	acceptedMeta.RewardEarned = make([]map[common.Hash]uint64, mutualLen)
+	for i := 0; i < mutualLen; i++ {
+		acceptedMeta.RewardEarned[i] = make(map[common.Hash]uint64)
+	}
+
+	if feeInPRV {
+		// uniformly split fee into reserves
+		feePerReserve := fee / uint64(mutualLen)
```
Comment (dat-incognito-org) [inline blockchain/pdex/v2utils/reserve.go:256]: due to integer division, it can happen that `len*feePerReserve < fee`. Do we consider the gap as burned in this case ?

File: common/constants.go [hunk 4]
```diff
@@ -68,6 +68,8 @@ var (
 	PRVCoinName           = "PRV"   // To send PRV in custom token
 	ConfidentialAssetID   = Hash{5}
 	ConfidentialAssetName = "CA"
+	PDEXCoinID            = Hash{6}
+	PDExCoinName          = "PDEX"
```
Comment (dat-incognito-org) [inline common/constants.go:72]: for consistency, shouldn't this constant be `PDEXCoinName` (capital X) ?

File: metadata/inittokenrequest.go [hunk 5]
```diff
@@ -4,10 +4,11 @@ import (
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
+	"strconv"
```
Comment (dat-incognito-org) [inline metadata/inittokenrequest.go:7]: should remove this change. This file is unrelated to pdexv3.

File: metadata/pdexv3/withdraw_lp_fee_response.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		isMinted, mintCoin, assetID, err := tx.GetTxMintData()
+		if err != nil {
+			continue
+		}
+		if !isMinted {
+			continue
+		}
+		pk := mintCoin.GetPublicKey().ToBytesS()
+		paidAmount := mintCoin.GetValue()
+		txR := mintCoin.(*coin.CoinV2).GetTxRandom()
```
Comment (dat-incognito-org) [inline metadata/pdexv3/withdraw_lp_fee_response.go:128]: need a comment explaining why the cast is guaranteed to succeed

File: metadata/pdexv3/withdraw_protocol_fee_response.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+		isMinted, mintCoin, assetID, err := tx.GetTxMintData()
+		if err != nil {
+			continue
+		}
+		if !isMinted {
+			continue
+		}
+		pk := mintCoin.GetPublicKey().ToBytesS()
+		paidAmount := mintCoin.GetValue()
+		txR := mintCoin.(*coin.CoinV2).GetTxRandom()
```
Comment (dat-incognito-org) [inline metadata/pdexv3/withdraw_protocol_fee_response.go:128]: same as above, need comment

File: blockchain/pdex/v2utils/producer.go (hunk 2 above)
Comment (cuong-incognito) [reply blockchain/pdex/v2utils/producer.go:142]: yes, I used it for appending keys to a slice, and after the loop of  `i++`, `i` doesn't be used anymore. I will rewrite it to make it more understandable.

File: blockchain/pdex/v2utils/reserve.go (hunk 3 above)
Comment (cuong-incognito) [reply blockchain/pdex/v2utils/reserve.go:256]: The maximum fee that is lost is the `MaxLengthPath -1` nano token. I think it is acceptable.

File: common/constants.go (hunk 4 above)
Comment (cuong-incognito) [reply common/constants.go:72]: Typo, Thank!

File: metadata/inittokenrequest.go (hunk 5 above)
Comment (cuong-incognito) [reply metadata/inittokenrequest.go:7]: my IDE automatically formatted it. Will undo this change.

File: blockchain/pdex/v2utils/reserve.go (hunk 3 above)
Comment (cuong-incognito) [reply blockchain/pdex/v2utils/reserve.go:256]: I think I will add the gap to the last pool.

File: metadata/pdexv3/withdraw_lp_fee_response.go (hunk 6 above)
Comment (cuong-incognito) [reply metadata/pdexv3/withdraw_lp_fee_response.go:128]: Copy from [here](https://github.com/incognitochain/incognito-chain/blob/69580c10ec2def108e28fafb52449b005a898437/metadata/pdecrosspooltraderesponse.go#L152). I think it always works after the privacy v2 checkpoint because all minting Txs create coin v2.

File: dataaccessobject/rawdbv2/accessor_pdexv3.go [hunk 8]
```diff
@@ -146,10 +146,9 @@ type Pdexv3PoolPair struct {
 	token0VirtualAmount *big.Int
 	token1VirtualAmount *big.Int
 	amplifier           uint
-}
-
-func (pp *Pdexv3PoolPair) SetShareAmount(shareAmount uint64) {
-	pp.shareAmount = shareAmount
+	lpFeesPerShare      map[common.Hash]*big.Int
```
Comment (tin-incognito) [inline dataaccessobject/rawdbv2/accessor_pdexv3.go:149]: This property gonna update with everytime trading fee increase right?

File: dataaccessobject/rawdbv2/accessor_pdexv3.go [hunk 9]
```diff
@@ -146,10 +146,9 @@ type Pdexv3PoolPair struct {
...
+	protocolFees        map[common.Hash]uint64
```
Comment (tin-incognito) [inline dataaccessobject/rawdbv2/accessor_pdexv3.go:150]: This property gonna update with everytime mint Pdex token right?

File: dataaccessobject/rawdbv2/accessor_pdexv3.go [hunk 10]
```diff
@@ -146,10 +146,9 @@ type Pdexv3PoolPair struct {
...
+	stakingPoolFees     map[common.Hash]uint64
```
Comment (tin-incognito) [inline dataaccessobject/rawdbv2/accessor_pdexv3.go:151]: When will this property gonna update?

File: dataaccessobject/statedb/stateobject_pdexv3_share.go [hunk 11]
```diff
@@ -3,15 +3,17 @@ package statedb
...
 	"github.com/incognitochain/incognito-chain/common"
 )
 
 type Pdexv3ShareState struct {
-	nftID                   common.Hash
-	amount                  uint64
-	lastUpdatedBeaconHeight uint64
+	nftID              common.Hash
+	amount             uint64
+	tradingFees        map[common.Hash]uint64
```
Comment (tin-incognito) [inline dataaccessobject/statedb/stateobject_pdexv3_share.go:15]: Each times trading fees increase, all struct `Pdexv3ShareState` gonna be update and store to new leaf in merle tree. That's why I separate it to `Pdexv3TradingFeeState`. You can refer it for optimizing pdexv3 state storage

File: metadata/pdexv3/mint_pdex_token.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+	"github.com/incognitochain/incognito-chain/common"
+	"github.com/incognitochain/incognito-chain/dataaccessobject/statedb"
+	metadataCommon "github.com/incognitochain/incognito-chain/metadata/common"
+	"github.com/incognitochain/incognito-chain/wallet"
+)
+
+type MintPDEXGenesisResponse struct {
+	metadataCommon.MetadataBase
+	MintingPaymentAddress string `json:"MintingPaymentAddress"`
```
Comment (tin-incognito) [inline metadata/pdexv3/mint_pdex_token.go:16]: paymentAddress?

Comment (cuong-incognito) [reply metadata/pdexv3/mint_pdex_token.go:16]: yes, this is the payment address of the DAO account, no need to censor this address

File: dataaccessobject/statedb/stateobject_pdexv3_share.go (hunk 11 above)
Comment (cuong-incognito) [reply dataaccessobject/statedb/stateobject_pdexv3_share.go:15]: the tradingFees will be updated only when the share amount is updated (or fee withdrawal) so I don't think we need another state object for only store 4 properties that won't be regularly updated.

File: blockchain/pdex/params.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+type Params struct {
+	DefaultFeeRateBPS               uint            // the default value if fee rate is not specific in FeeRateBPS (default 0.3% ~ 30 BPS)
+	FeeRateBPS                      map[string]uint // map: pool ID -> fee rate (0.1% ~ 10 BPS)
+	PRVDiscountPercent              uint            // percent of fee that will be discounted if using PRV as the trading token fee (default: 25%)
+	TradingProtocolFeePercent       uint            // percent of fees that is rewarded for the core team (default: 0%)
+	TradingStakingPoolRewardPercent uint            // percent of fees that is distributed for staking pools (PRV, PDEX, ..., default: 10%)
+	PDEXRewardPoolPairsShare        map[string]uint // map: pool pair ID -> PDEX reward share weight
+	StakingPoolsShare               map[string]uint // map: staking tokenID -> pool staking share weight
+	StakingRewardTokens             []common.Hash   // list of staking reward tokens
```
[the remark this answers is not in the document]
Comment (cuong-incognito) [reply blockchain/pdex/params.go:18]: map of `StakingRewardTokens`?

File: dataaccessobject/rawdbv2/accessor_pdexv3.go (hunk 9 above)
Comment (cuong-incognito) [reply dataaccessobject/rawdbv2/accessor_pdexv3.go:150]: no, protocolFees only updated when a trade is executed and the percent of protocol fee is larger than 0.
Minting PDEX token in genesis block won't be stored in the StateDB.

File: dataaccessobject/rawdbv2/accessor_pdexv3.go (hunk 8 above)
Comment (cuong-incognito) [reply dataaccessobject/rawdbv2/accessor_pdexv3.go:149]: yes

File: dataaccessobject/rawdbv2/accessor_pdexv3.go (hunk 10 above)
Comment (cuong-incognito) [reply dataaccessobject/rawdbv2/accessor_pdexv3.go:151]: same as the protocol fee

File: blockchain/pdex/pool_pair_state.go [hunk 14]
```diff
@@ -344,6 +360,52 @@ func (p *PoolPairState) updateSingleTokenAmount(
...
+	curShare, ok := p.shares[nftID.String()]
+	if !ok {
+		return nil, fmt.Errorf("Share not found")
+	}
+
+	curLPFeesPerShare := p.state.LPFeesPerShare()
+	oldLPFeesPerShare := curShare.lastLPFeesPerShare
+
+	listTokenIDs := []common.Hash{
+		p.state.Token0ID(),
```
[the remark this answers is not in the document]
Comment (cuong-incognito) [reply blockchain/pdex/pool_pair_state.go:377]: duplicate? can you give me some details?

File: metadata/pdexv3/withdraw_lp_fee_response.go (hunk 6 above)
Comment (dat-incognito-org) [reply metadata/pdexv3/withdraw_lp_fee_response.go:128]: explain there, don't explain here
one cannot tell this always succeeds without looking at a lot of other code, which is why it needs explanation

File: blockchain/pdex/params.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tin-incognito) [reply blockchain/pdex/params.go:18]: yes O(1) map[string]bool

File: metadata/pdexv3/withdraw_lp_fee_response.go (hunk 6 above)
Comment (cuong-incognito) [reply metadata/pdexv3/withdraw_lp_fee_response.go:128]: You miss the latter part of my comment. For safety reasons, I will have a type casting here just for sure.

File: blockchain/pdex/pool_pair_state.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:377]: listTokenIDs := append(p.state.Token0ID(), p.state.Token1ID())
listTokenIDs = append(listTokenIDs, availableListTokenIDs)

File: blockchain/pdex/interface.go [hunk 15]
```diff
@@ -14,7 +14,7 @@ type State interface {
 }
 
 type StateReader interface {
-	Params() Params
+	Params() *Params
```
[the remark this answers is not in the document]
Comment (cuong-incognito) [reply blockchain/pdex/interface.go:17]: yes

File: blockchain/pdex/params.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (cuong-incognito) [reply blockchain/pdex/params.go:18]: I don't think this list won't be regularly updated. So a list instead of a map is fine by me.

[the remark this answers is not in the document]
Comment (tin-incognito) [reply blockchain/pdex/params.go:18]: It help for ur handler in mem to acceess more quickly to data u want in mem. But it's ok if the length of StakingRewardsTokens is not so large

File: blockchain/pdex/pool_pair_state.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (cuong-incognito) [reply blockchain/pdex/pool_pair_state.go:377]: I doubt it will help the codes more clear

File: metadata/pdexv3/withdraw_protocol_fee_response.go (hunk 7 above)
Comment (cuong-incognito) [reply metadata/pdexv3/withdraw_protocol_fee_response.go]: fixed

File: metadata/pdexv3/withdraw_lp_fee_response.go (hunk 6 above)
Comment (cuong-incognito) [reply metadata/pdexv3/withdraw_lp_fee_response.go]: fixed

File: blockchain/pdex/v2utils/reserve.go (hunk 3 above)
Comment (cuong-incognito) [reply blockchain/pdex/v2utils/reserve.go:256]: fix

File: blockchain/pdex/params.go (hunk 1 above)
Comment (cuong-incognito) [reply blockchain/pdex/params.go:39]: fixed

**Review by dat-incognito-org — APPROVED** [review]

**Review by tin-incognito — APPROVED** [review]
LGTM!
